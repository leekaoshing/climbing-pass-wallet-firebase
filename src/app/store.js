import { configureStore } from '@reduxjs/toolkit';
import { firebaseReducer } from 'react-redux-firebase';
import { actionTypes, firestoreReducer } from 'redux-firestore';
import userReducer from '../reducers/userSlice';
import dialogReducer from '../reducers/dialogSlice';

// import { createStore, applyMiddleware, compose } from 'redux';
// import reduxThunk from 'redux-thunk';
// import { ReactReduxFirebaseProvider, getFirebase } from 'react-redux-firebase'
// import { reduxFirestore, getFirestore } from 'redux-firestore';
// import { devToolsEnhancer, composeWithDevTools } from 'redux-devtools-extension';
// import thunk from 'redux-thunk';
// import firebase from '../services/firebase';

export const rrfConfig = {
  userProfile: 'users',
  attachAuthIsReady: true,
  useFirestoreForProfile: true,
  onAuthStateChanged: (authData, firebase, dispatch) => {
    // Clear redux-firestore state if auth does not exist (i.e logout)
    if (!authData) {
      dispatch({ type: actionTypes.CLEAR_DATA })
    }
 }
}

// export default configureStore = () => {
//   const middlewares = [thunk.withExtraArgument(getFirebase)];

//   const composedEnhancer = composeWithDevTools(
//     applyMiddleware(...middlewares),
//     reactReduxFirebase(firebase, rrfConfig),
//     reduxFirestore(firebase)
//   );

//   const store = createStore({
//     user: userReducer,
//     gym: gymReducer
//   }, composedEnhancer)

//   return store
// }

// const rootReducer = combineReducers({
//   user: userReducer,
//   gym: gymReducer,
//   firebase: firebaseReducer,
//   firestore: firestoreReducer
// })

// // Create store with reducers and initial state
// const initialState = {}
// export default createStore(rootReducer, initialState);

export const rootReducer = {
  user: userReducer,
  dialog: dialogReducer,
  firebase: firebaseReducer,
  firestore: firestoreReducer
};

export default configureStore({
  reducer: rootReducer
});
