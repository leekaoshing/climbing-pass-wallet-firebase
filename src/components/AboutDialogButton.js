import Button from '@material-ui/core/Button';
import React from "react";
import { useDispatch } from 'react-redux';
import {
    setShowAboutDialog
} from '../reducers/dialogSlice';

export function AboutDialogButton() {
    const dispatch = useDispatch();

    const handleOpen = () => {
        dispatch(setShowAboutDialog(true));
    }

    return (
        <Button
            color="inherit"
            variant="outlined"
            size="small"
            onClick={handleOpen}
        >
            About
        </Button>
    );
};