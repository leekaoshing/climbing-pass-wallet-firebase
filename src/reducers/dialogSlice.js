import { createSlice } from '@reduxjs/toolkit';

export const dialogSlice = createSlice({
  name: 'dialog',
  initialState: {
    showAboutDialog: false,
    showAddGymDialog: false,
    showConfirmationDialog: false,
    showCreateUserDialog: false,
    showPasswordResetDialog: false,
    showUpdateResultDialog: false,
  },
  reducers: {
    setShowAboutDialog: (state, action) => {
      state.showAboutDialog = action.payload;
    },
    setShowAddGymDialog: (state, action) => {
      state.showAddGymDialog = action.payload;
    },
    setShowConfirmationDialog: (state, action) => {
      state.showConfirmationDialog = action.payload;
    },
    setShowCreateUserDialog: (state, action) => {
      state.showCreateUserDialog = action.payload;
    },
    setShowPasswordResetDialog: (state, action) => {
      state.showPasswordResetDialog = action.payload;
    },
    setShowUpdateResultDialog: (state, action) => {
      state.showUpdateResultDialog = action.payload;
    }
  }
});

export const {
  setShowAboutDialog,
  setShowAddGymDialog,
  setShowConfirmationDialog,
  setShowCreateUserDialog,
  setShowPasswordResetDialog,
  setShowUpdateResultDialog,
} = dialogSlice.actions;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectShowAboutDialog = state => {
  return state.dialog.showAboutDialog;
}

export const selectShowAddGymDialog = state => {
  return state.dialog.showAddGymDialog;
}

export const selectShowConfirmationDialog = state => {
  return state.dialog.showConfirmationDialog;
}

export const selectShowCreateUserDialog = state => {
  return state.dialog.showCreateUserDialog;
}

export const selectShowPasswordResetDialog = state => {
  return state.dialog.showPasswordResetDialog;
}

export const selectShowUpdateResultDialog = state => {
  return state.dialog.showUpdateResultDialog;
}

export default dialogSlice.reducer;
