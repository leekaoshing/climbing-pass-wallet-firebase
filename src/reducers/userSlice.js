import { createSlice } from '@reduxjs/toolkit';

export const userSlice = createSlice({
  name: 'user',
  initialState: {
    editableUser: null,
    loadingUpdateUser: false,
    updateResult: null,
  },
  reducers: {
    addGymToEditableUser: (state, action) => {
      state.editableUser.passes[action.payload] = 0
    },
    setEditableUser: (state, action) => {
      state.editableUser = action.payload;
    },
    setLoadingUpdateUser: (state, action) => {
      state.loadingUpdateUser = action.payload;
    },
    setUpdateResult: (state, action) => {
      state.updateResult = action.payload;
    }
  }
});

export const {
  addGymToEditableUser,
  setEditableUser,
  setLoadingUpdateUser,
  setUpdateResult
} = userSlice.actions;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectEditableUser = state => {
  return state.user.editableUser;
}

export const selectLoadingUpdateUser = state => {
  return state.user.loadingUpdateUser;
}

export const selectUpdateResult = state => {
  return state.user.updateResult;
}

export const getPassDifferences = state => {
  const auth = state.firebase.auth;
  const oldPasses = state.firestore.data.users[auth.uid].passes;
  const newPasses = state.user.editableUser ? state.user.editableUser.passes : {};

  const differenceInPasses = {};
  Object.keys(newPasses).forEach(key => {
    const oldPassQuantity = oldPasses[key] === undefined ? 0 : oldPasses[key];
    const newPassQuantity = newPasses[key];
    const change = newPassQuantity - oldPassQuantity;
    if (change !== 0) differenceInPasses[key] = newPassQuantity - oldPassQuantity;
  })
  return differenceInPasses;
}

export default userSlice.reducer;
