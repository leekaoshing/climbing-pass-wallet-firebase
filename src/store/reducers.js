import { combineReducers } from 'redux'
import firebase from 'react-redux-firebase/lib/reducer'
import firestore from 'redux-firestore/lib/reducer'
import locationReducer from './reducers/location'
import dialogReducer from './reducers/dialog'
import userReducer from './reducers/user'

export function makeRootReducer(asyncReducers) {
  return combineReducers({
    // Add sync reducers here
    firebase,
    firestore,
    dialog: dialogReducer,
    user: userReducer,
    location: locationReducer,
    ...asyncReducers
  })
}

export function injectReducer(store, { key, reducer }) {
  store.asyncReducers[key] = reducer // eslint-disable-line no-param-reassign
  store.replaceReducer(makeRootReducer(store.asyncReducers))
}

export default makeRootReducer
