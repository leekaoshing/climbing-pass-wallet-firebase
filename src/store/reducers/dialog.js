import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  showUpdateResultDialog: false,
}

export const dialogSlice = createSlice({
  name: 'dialog',
  initialState,
  reducers: {
    setShowUpdateResultDialog: (state, action) => {
      state.showUpdateResultDialog = action.payload
    },
    resetState: () => initialState,
  }
})

export const {
  resetState,
  setShowUpdateResultDialog
} = dialogSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectShowUpdateResultDialog = state => {
  return state.dialog.showUpdateResultDialog
}

export default dialogSlice.reducer
