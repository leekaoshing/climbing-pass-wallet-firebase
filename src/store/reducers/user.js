import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  passes: null,
  updateResult: null
}

export const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    addGymToUser: (state, action) => {
      state.passes[action.payload] = 0
    },
    setPassesForGym: (state, action) => {
      const gymId = action.payload.gymId
      const numberOfPasses = action.payload.numberOfPasses
      state.passes[gymId] = numberOfPasses
    },
    setPassesForUser: (state, action) => {
      state.passes = action.payload
    },
    setUpdateResult: (state, action) => {
      state.updateResult = action.payload
    },
    resetState: () => initialState,
  }
})

export const {
  addGymToUser,
  resetState,
  setPassesForGym,
  setPassesForUser,
  setUpdateResult
} = userSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectUserPasses = state => {
  return state.user.passes
}

export const selectUpdateResult = state => {
  return state.user.updateResult
}

export default userSlice.reducer
