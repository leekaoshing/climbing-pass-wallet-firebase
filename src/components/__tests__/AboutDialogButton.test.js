import {
    fireEvent, render,
    screen
} from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../app/store';
import { selectShowAboutDialog } from '../../reducers/dialogSlice';
import { AboutDialogButton } from '../AboutDialogButton';

jest.disableAutomock()

describe.skip('About Dialog Button', () => {
    it('renders and performs show dialog action correctly', () => {
        const { asFragment } = render(
            <Provider store={store}>
                <AboutDialogButton />
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
        const button = screen.getByText(/About/i);
        fireEvent.click(button);

        expect(selectShowAboutDialog(store.getState())).toBe(true);
    });
})