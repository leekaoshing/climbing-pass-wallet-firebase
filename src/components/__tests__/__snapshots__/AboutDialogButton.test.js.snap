// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About Dialog Button renders and performs show dialog action correctly 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-colorInherit MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      About
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
